extensions/notification/NotificationQueuePopulator.js

```diff
@@ -119,22 +119,78 @@ class NotificationQueuePopulator extends QueuePopulatorExtension {
         return this.bnConfigManager.removeConfig(bucketName);
     }
 
+    /**
+     * Returns the correct versionId
+     * to display according to the
+     * versioning state of the object
+     * @param {Object} value log entry object
+     * @param {Object} overheadFields - extra fields missing from the log entry
+     * @return {String} versionId
+     */
+    _getVersionId(value, overheadFields) {
+        const versionId = value.versionId || (overheadFields && overheadFields.versionId);
+        const isNullVersion = value.isNull;
+        const isVersioned = !!versionId;
+        // Versioning suspended objects have
+        // a versionId, however it is internal
+        // and should not be used to get the object
+        if (isNullVersion || !isVersioned) {
+            return null;
+        }
+        return versionId;
+    }
+
+    /**
+     * Decides if we should process the entry.
+     * Since we get both master and version events,
+     * we need to avoid pushing two notifications for
+     * the same event.
+     * - For non versiond buckets, we process the master
+     * objects' events.
+     * - For versioned buckets, we process version events
+     * and ignore all master events.
+     * - For versioning suspended buckets, we need to process
+     * both master and version events, as the master is considered
+     * a separate version.
+     * @param {String} key object key
+     * @param {Object} value object metadata
+     * @return {boolean} - true if entry is valid
+     */
+    _shouldProcessEntry(key, value) {
+        const isMaster = isMasterKey(key);
+        const isVersionedMaster = isMaster && !!value.versionId;
+        const isNullVersion = isMaster && value.isNull;
```

Comment:
Beware that in V0 format, null versions may have both a master and a versioned key. This is due to a design flaw, it has been fixed in the rework of the null versions format for V1 (and for V0 when V1 is enabled). For this reason, I think we should also exclude versioned keys that have the `isNull` attribute.

But... in V1, null versions may **only** be in a versioned key, however this version key will then be a null key (i.e. its version ID is the empty string, or the raw key ends with `\0`).

So, I think to be correct we should, in addition to this logic, exclude versioned keys (with a non-empty version ID) that have the `isNull` attribute. Null keys are guaranteed not to be duplicated with a master key in the new logic.

Reply:
@Kerkesni how do we handle this in backbeat 8 ? (if we actually do handle it...)

Reply:
"null" **keys** (those ending with `\0`) have only been implemented in S3C, Artesca uses master or versioned keys for null versions.